File: src/main/java/com/y271727uy/ranch_festival/dimension/DimensionDefinition.java
package com.y271727uy.ranch_festival.dimension;

import com.y271727uy.ranch_festival.RanchFestivalMod;
import com.y271727uy.ranch_festival.all.ModItem;
import java.util.Collections;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Objects;
import java.util.Set;
import java.util.function.Supplier;
import net.minecraft.core.BlockPos;
import net.minecraft.core.registries.Registries;
import net.minecraft.resources.ResourceKey;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.Level;
import net.minecraftforge.registries.RegistryObject;
import sereneseasons.api.season.Season;

@SuppressWarnings("unused")
public final class DimensionDefinition {
    private static final Set<DimensionDefinition> DEFINITIONS = new LinkedHashSet<>();

    public static final DimensionDefinition TEST = create()
            .Structure((ResourceLocation) null)
            .StructureAxis(0, 64, 0)
            .StructureDimension(ResourceKey.create(
                    Registries.DIMENSION,
                    ResourceLocation.fromNamespaceAndPath(RanchFestivalMod.MODID, "test")
            ))
            .PlayerInputAxis(0, 64, 0)
            .Seasons(Season.SPRING)
            .SubSeason(null)
            .Day(2)
            .SurvivalPlayerDestory(false)
            .AutoReset(true)
            .MobSpawn(false)
            .invitation(ModItem.TEXT_INVITATION)
            .register();

    private final ResourceLocation structure;
    private final BlockPos structureAxis;
    private final ResourceKey<Level> structureDimension;
    private final BlockPos playerInputAxis;
    private final Set<Season> seasons;
    private final Season.SubSeason subSeason;
    private final Integer day;
    private final Boolean survivalPlayerDestroy;
    private final Boolean autoReset;
    private final Boolean mobSpawn;
    private final Supplier<? extends Item> invitation;

    private DimensionDefinition(Builder builder) {
        this.structure = builder.structure;
        this.structureAxis = builder.structureAxis;
        this.structureDimension = builder.structureDimension;
        this.playerInputAxis = builder.playerInputAxis;
        this.seasons = builder.seasons;
        this.subSeason = builder.subSeason;
        this.day = builder.day;
        this.survivalPlayerDestroy = builder.survivalPlayerDestroy;
        this.autoReset = builder.autoReset;
        this.mobSpawn = builder.mobSpawn;
        this.invitation = builder.invitation;
    }

    public static Builder create() {
        return new Builder();
    }

    public static DimensionDefinition register(Builder builder) {
        return register(builder.build());
    }

    public static DimensionDefinition register(DimensionDefinition definition) {
        DEFINITIONS.add(definition);
        return definition;
    }

    public static DimensionDefinition findByDimension(ResourceKey<Level> dimension) {
        for (DimensionDefinition definition : DEFINITIONS) {
            if (Objects.equals(definition.getStructureDimension(), dimension)) {
                return definition;
            }
        }
        return null;
    }

    public static DimensionDefinition findByInvitation(ItemStack itemStack) {
        if (itemStack == null || itemStack.isEmpty()) {
            return null;
        }

        Item item = itemStack.getItem();
        for (DimensionDefinition definition : DEFINITIONS) {
            if (definition.matchesInvitation(item)) {
                return definition;
            }
        }
        return null;
    }

    public ResourceLocation getStructure() {
        return structure;
    }

    public BlockPos getStructureAxis() {
        return structureAxis;
    }

    public ResourceKey<Level> getStructureDimension() {
        return structureDimension;
    }

    public BlockPos getPlayerInputAxis() {
        return playerInputAxis;
    }

    public Set<Season> getSeasons() {
        return seasons;
    }

    public Season.SubSeason getSubSeason() {
        return subSeason;
    }

    public Integer getDay() {
        return day;
    }

    public Boolean getSurvivalPlayerDestroy() {
        return survivalPlayerDestroy;
    }

    public Boolean getAutoReset() {
        return autoReset;
    }

    public Boolean getMobSpawn() {
        return mobSpawn;
    }

    public Item getInvitationItem() {
        return invitation == null ? null : invitation.get();
    }

    public boolean matchesInvitation(Item item) {
        return invitation != null && item != null && getInvitationItem() == item;
    }

    public boolean matchesSeason(Season season) {
        return seasons == null || seasons.isEmpty() || (season != null && seasons.contains(season));
    }

    public boolean matchesSubSeason(Season.SubSeason currentSubSeason) {
        return subSeason == null || Objects.equals(subSeason, currentSubSeason);
    }

    public boolean matchesDay(int currentDay) {
        return day == null || day == currentDay;
    }

    public static final class Builder {
        private ResourceLocation structure;
        private BlockPos structureAxis;
        private ResourceKey<Level> structureDimension;
        private BlockPos playerInputAxis;
        private Set<Season> seasons;
        private Season.SubSeason subSeason;
        private Integer day;
        private Boolean survivalPlayerDestroy;
        private Boolean autoReset;
        private Boolean mobSpawn;
        private Supplier<? extends Item> invitation;

        private Builder() {
        }

        public Builder Structure(String structurePath) {
            this.structure = structurePath == null ? null : ResourceLocation.fromNamespaceAndPath(RanchFestivalMod.MODID, structurePath);
            return this;
        }

        public Builder Structure(ResourceLocation structure) {
            this.structure = structure;
            return this;
        }

        public Builder StructureAxis(int x, int y, int z) {
            this.structureAxis = new BlockPos(x, y, z);
            return this;
        }

        public Builder StructureAxis(BlockPos structureAxis) {
            this.structureAxis = structureAxis;
            return this;
        }

        public Builder StructureDimension(ResourceKey<Level> structureDimension) {
            this.structureDimension = structureDimension;
            return this;
        }

        public Builder PlayerInputAxis(int x, int y, int z) {
            this.playerInputAxis = new BlockPos(x, y, z);
            return this;
        }

        public Builder PlayerInputAxis(BlockPos playerInputAxis) {
            this.playerInputAxis = playerInputAxis;
            return this;
        }

        public Builder Seasons(Season... seasons) {
            if (seasons == null || seasons.length == 0) {
                this.seasons = null;
                return this;
            }

            Set<Season> configuredSeasons = new LinkedHashSet<>();
            Collections.addAll(configuredSeasons, seasons);
            this.seasons = Collections.unmodifiableSet(configuredSeasons);
            return this;
        }

        public Builder SubSeason(Season.SubSeason subSeason) {
            this.subSeason = subSeason;
            return this;
        }

        public Builder Day(Integer day) {
            this.day = day;
            return this;
        }

        public Builder SurvivalPlayerDestory(Boolean survivalPlayerDestroy) {
            this.survivalPlayerDestroy = survivalPlayerDestroy;
            return this;
        }

        public Builder SurvivalPlayerDestroy(Boolean survivalPlayerDestroy) {
            return SurvivalPlayerDestory(survivalPlayerDestroy);
        }

        public Builder AutoReset(Boolean autoReset) {
            this.autoReset = autoReset;
            return this;
        }

        public Builder MobSpawn(Boolean mobSpawn) {
            this.mobSpawn = mobSpawn;
            return this;
        }

        public Builder invitation(Item invitation) {
            return invitation(invitation == null ? null : () -> invitation);
        }

        public Builder invitation(RegistryObject<? extends Item> invitation) {
            return invitation(invitation == null ? null : () -> invitation.get());
        }

        public Builder invitation(Supplier<? extends Item> invitation) {
            this.invitation = invitation;
            return this;
        }

        public DimensionDefinition build() {
            return new DimensionDefinition(this);
        }

        public DimensionDefinition register() {
            return DimensionDefinition.register(build());
        }
    }
}

File: src/main/kotlin/com/y271727uy/ranch_festival/dimension/DimensionDefinition.kt
package com.y271727uy.ranch_festival.dimension

import com.y271727uy.ranch_festival.RanchFestivalMod
import java.util.Collections
import java.util.LinkedHashSet
import java.util.Objects
import java.util.function.Supplier
import net.minecraft.core.BlockPos
import net.minecraft.network.chat.Component
import net.minecraft.resources.ResourceKey
import net.minecraft.resources.ResourceLocation
import net.minecraft.server.level.ServerLevel
import net.minecraft.world.item.Item
import net.minecraft.world.level.Level
import net.minecraftforge.registries.RegistryObject
import sereneseasons.api.season.Season

class DimensionDefinition private constructor(builder: Builder) {
    val structure: ResourceLocation? = builder.structureValue
    val structureAxis: BlockPos? = builder.structureAxisValue
    val structureDimension: ResourceKey<Level>? = builder.structureDimensionValue
    val playerInputAxis: BlockPos? = builder.playerInputAxisValue
    val seasons: Set<Season>? = builder.seasonsValue
    val subSeason: Season.SubSeason? = builder.subSeasonValue
    val day: Int? = builder.dayValue
    val dimensionNameKey: String? = builder.dimensionNameKeyValue
    val festivalNameKey: String? = builder.festivalNameKeyValue
    val survivalPlayerDestroy: Boolean? = builder.survivalPlayerDestroyValue
    val autoReset: Boolean? = builder.autoResetValue
    val mobSpawn: Boolean? = builder.mobSpawnValue
    private val invitation: Supplier<out Item>? = builder.invitationValue

    fun getInvitationItem(): Item? = invitation?.get()

    fun matchesInvitation(item: Item?): Boolean {
        return invitation != null && item != null && getInvitationItem() === item
    }

    fun matchesSeason(season: Season?): Boolean {
        return seasons.isNullOrEmpty() || (season != null && seasons.contains(season))
    }

    fun matchesSubSeason(currentSubSeason: Season.SubSeason?): Boolean {
        return subSeason == null || Objects.equals(subSeason, currentSubSeason)
    }

    fun matchesDay(currentDay: Int): Boolean {
        return day == null || day == currentDay
    }

    @Suppress("unused")
    fun getDimensionDisplayName(): Component {
        val key = dimensionNameKey ?: festivalNameKey
        if (!key.isNullOrBlank()) {
            return Component.translatable(key)
        }

        val dimensionId = structureDimension?.location()?.toString()
        return if (!dimensionId.isNullOrBlank()) {
            Component.literal(dimensionId)
        } else {
            Component.translatable("ranch_festival.message.unknown_dimension")
        }
    }

    fun resolveStructureClearBounds(level: ServerLevel, padding: Int = 3): StructureClearBounds? {
        val origin = structureAxis ?: return null
        val structureId = structure ?: return null
        val template = level.structureManager.getOrCreate(structureId)
        val size = template.size
        val extra = padding.coerceAtLeast(0)
        val farCorner = origin.offset(size.x - 1, size.y - 1, size.z - 1)

        return StructureClearBounds(
            BlockPos(
                minOf(origin.x, farCorner.x) - extra,
                minOf(origin.y, farCorner.y) - extra,
                minOf(origin.z, farCorner.z) - extra
            ),
            BlockPos(
                maxOf(origin.x, farCorner.x) + extra,
                maxOf(origin.y, farCorner.y) + extra,
                maxOf(origin.z, farCorner.z) + extra
            )
        )
    }

    companion object {
        @JvmStatic
        fun create(): Builder = Builder()
    }

    @Suppress("unused")
    class Builder {
        var structureValue: ResourceLocation? = null
        var structureAxisValue: BlockPos? = null
        var structureDimensionValue: ResourceKey<Level>? = null
        var playerInputAxisValue: BlockPos? = null
        var seasonsValue: Set<Season>? = null
        var subSeasonValue: Season.SubSeason? = null
        var dayValue: Int? = null
        var dimensionNameKeyValue: String? = null
        var festivalNameKeyValue: String? = null
        var survivalPlayerDestroyValue: Boolean? = null
        var autoResetValue: Boolean? = null
        var mobSpawnValue: Boolean? = null
        var invitationValue: Supplier<out Item>? = null

        fun structure(structurePath: String?): Builder = apply {
            structureValue = structurePath?.let { ResourceLocation.fromNamespaceAndPath(RanchFestivalMod.MODID, it) }
        }

        fun structure(structure: ResourceLocation?): Builder = apply {
            structureValue = structure
        }

        fun structureAxis(x: Int, y: Int, z: Int): Builder = apply {
            structureAxisValue = BlockPos(x, y, z)
        }

        fun structureAxis(structureAxis: BlockPos?): Builder = apply {
            structureAxisValue = structureAxis
        }

        fun structureDimension(structureDimension: ResourceKey<Level>?): Builder = apply {
            structureDimensionValue = structureDimension
        }

        fun playerInputAxis(x: Int, y: Int, z: Int): Builder = apply {
            playerInputAxisValue = BlockPos(x, y, z)
        }

        fun playerInputAxis(playerInputAxis: BlockPos?): Builder = apply {
            playerInputAxisValue = playerInputAxis
        }

        fun seasons(vararg seasons: Season?): Builder {
            if (seasons.isEmpty()) {
                seasonsValue = null
                return this
            }

            val configuredSeasons = LinkedHashSet<Season>()
            seasons.filterNotNull().forEach(configuredSeasons::add)
            seasonsValue = if (configuredSeasons.isEmpty()) null else Collections.unmodifiableSet(configuredSeasons)
            return this
        }

        fun subSeason(subSeason: Season.SubSeason?): Builder = apply {
            subSeasonValue = subSeason
        }

        fun day(day: Int?): Builder = apply {
            dayValue = day
        }

        fun dimensionName(dimensionNameKey: String?): Builder = apply {
            dimensionNameKeyValue = dimensionNameKey
        }

        fun dimensionNameKey(dimensionNameKey: String?): Builder = dimensionName(dimensionNameKey)

        fun festivalName(festivalNameKey: String?): Builder = apply {
            festivalNameKeyValue = festivalNameKey
        }

        fun festivalNameKey(festivalNameKey: String?): Builder = festivalName(festivalNameKey)

        fun survivalPlayerDestory(survivalPlayerDestroy: Boolean?): Builder = apply {
            survivalPlayerDestroyValue = survivalPlayerDestroy
        }

        fun survivalPlayerDestroy(survivalPlayerDestroy: Boolean?): Builder = survivalPlayerDestory(survivalPlayerDestroy)

        fun autoReset(autoReset: Boolean?): Builder = apply {
            autoResetValue = autoReset
        }

        fun mobSpawn(mobSpawn: Boolean?): Builder = apply {
            mobSpawnValue = mobSpawn
        }

        fun invitation(invitation: Item?): Builder = invitation(invitation?.let { Supplier { it } })

        fun invitation(invitation: RegistryObject<out Item>?): Builder = invitation(invitation?.let { Supplier { it.get() } })

        fun invitation(invitation: Supplier<out Item>?): Builder = apply {
            invitationValue = invitation
        }

        fun build(): DimensionDefinition = DimensionDefinition(this)

        fun register(): DimensionDefinition = DimensionRegistry.register(build())
    }
}

fun dimensionDefinition(block: DimensionDefinition.Builder.() -> Unit): DimensionDefinition.Builder {
    return DimensionDefinition.create().apply(block)
}

data class StructureClearBounds(val start: BlockPos, val end: BlockPos)
